perfkitbenchmarker/providers/openstack/os_virtual_machine.py

```diff
@@ -33,214 +43,327 @@
+    if self.zone in self.validated_resources_set:
+      return  # No need to check again
+    with self._lock:
+      if self.zone in self.validated_resources_set:
+        return
+      logging.info('Validating prerequisites.')
+      self._CheckImage()
+      self._CheckFlavor()
+      self._CheckNetworks()
+      self.validated_resources_set.add(self.zone)
+      logging.info('Prerequisites validated.')
+
+  def _CheckImage(self):
+    """Tries to get image, if found continues execution otherwise aborts."""
+    cmd = os_utils.OpenStackCLICommand(self, 'image', 'show', self.image)
+    stdout, stderr, _ = cmd.Issue()
+    if stderr:
+      raise errors.Config.InvalidValue(' '.join(
+          ('Image %s could not be found.' % self.image,
+           'For valid image IDs/names run "openstack image list".',)))
+
+  def _CheckFlavor(self):
+    """Tries to get flavor, if found continues execution otherwise aborts."""
+    cmd = os_utils.OpenStackCLICommand(self, 'flavor', 'show',
+                                       self.machine_type)
+    stdout, stderr, _ = cmd.Issue()
+    if stderr:
+      raise errors.Config.InvalidValue(' '.join(
+          ('Machine type %s could not be found.' % self.machine_type,
+           'For valid machine type IDs/names run "openstack flavor list".',)))
+
+  def _CheckNetworks(self):
+    """Tries to get network, if found continues execution otherwise aborts."""
+    if not self.network_name:
+      if self.floating_ip_pool_name:
+        msg = ('Cannot associate floating-ip address from pool %s without '
+               'an internally routable network. Make sure '
+               '--openstack_network flag is set.')
+        raise errors.Error(msg)
+      else:
+        msg = ('Cannot build instance without a network. Make sure to set '
+               'either just --openstack_network or both '
+               '--openstack_network and --openstack_floating_ip_pool flags.')
+        raise errors.Error(msg)
+
+    self._CheckNetworkExists()
+
+    if self.floating_ip_pool_name:
+      self._CheckFloatingIPNetworkExists()
+
+  def _CheckFloatingIPNetworkExists(self):
+    cmd = os_utils.OpenStackCLICommand(self, 'ip', 'floating', 'pool', 'list')
+    stdout, stderr, _ = cmd.Issue()
+    resp = json.loads(stdout)
+    for flip_pool in resp:
+      if flip_pool['Name'] == self.floating_ip_pool_name:
+        break
+    else:
+      msg = ' '.join(('Floating IP pool %s could not be found.'
+                      % self.floating_ip_pool_name,
+                      'For valid floating IP pools run',
+                      '"openstack ip floating pool list".',))
+      raise errors.Config.InvalidValue(msg)
+
+  def _CheckNetworkExists(self):
+    cmd = os_utils.OpenStackCLICommand(self, 'network', 'show',
+                                       self.network_name)
+    stdout, stderr, _ = cmd.Issue()
+    if stderr:
+      msg = ' '.join(('Network %s could not be found.' % self.network_name,
```

Comment:
`_CheckImage` and `_CheckFlavor` also use `join()`. The tricky bit is replacing the value, I guess another way to do it is using `format()`. Create the automatically appended string using `{n}` placeholders and assign it to `msg`, then call `msg.format(...)` with the values.

Unless you are suggesting to do everything inline like I did for `_CheckImage` and `_CheckFlavor` instead of assingning it to `msg`.